New metrics

mattevans opened the pull request:
This change removes `avg` as a metric and splits it out into three new metrics (`mean`, `median` and `mode`). Also implemented is `sum`, `cardinality` and `stdev`

mattevans committed:
Using const for repeating values

mattevans committed:
Added new metrics (mean, median, mode, stdev, sum, cardinality)

mattevans committed:
Updated tests for new metrics

snikch commented on metrics.go:
```diff
@@ -29,25 +41,104 @@ type measurer interface {
+		return nil
+	}
+
+	numbers := []float64{}
```
So a couple of questions about the mode.

1) Do we need to do two loops? Looks like you could cover this all in a single loop.
2) Why do we return an empty loop if the modes are all of the numbers / one? I haven't dealt with modes too much but wouldn't it still be relevant to have the only possible value returned regardless?

If you could add some comments re: mode that'd be good too - in fact we should quickly explain what each of the averages are - I always get them mixed up, so it'd be nice not to have to think too hard.

snikch commented on metrics.go:
```diff
@@ -91,3 +183,76 @@ func (a *max) Result() interface{} {
 	return result
 }
+
+// Cadinality
```
Spelling.

snikch commented on metrics.go:
```diff
@@ -91,3 +183,76 @@ func (a *max) Result() interface{} {
+}
+
+func (a *cardinality) AddDatum(datum interface{}) {
+	a.size++
```
Cardinality should be the total unique elements added, so this should really be a `map[interface{}]bool` where the values are comparable, and the `Result` should call `len(that_map)`.

snikch replied:
If you can't just use interface{} because it doesn't consider a == b, then perhaps make an actual interface

snikch replied:
Or you could just do a known type switch, i.e

```
switch t := datum.(type) {
  case string:
    a.values[t]++
…
}
```

snikch commented on metrics.go:
```diff
@@ -29,25 +41,104 @@ type measurer interface {
+	// Even slice? Get the mean of the middle values.
+	if len(a.list)%2 == 0 {
+		prev, _ := a.list[middle-1].Float64()
+		median = (median + prev) / 2
```
Might be worth doing decimal math here.

snikch commented on metrics.go:
```diff
@@ -29,25 +41,104 @@ type measurer interface {
+	sort.Sort(decimalSortNumerical(a.list))
+
+	// Determine median value.
+	middle := len(a.list) / 2
```
You could probably do this in an `else` below, rather than double working.

snikch commented on processor.go:
```diff
@@ -245,8 +246,22 @@ func (p *queryProcessor) measure() {
 				row := bucket.sourceRows[j]
+
+				// If measurer is cardinality, we can +1 not worrying about field value.
+				if (reflect.TypeOf(m) == reflect.TypeOf(&cardinality{})) {
```
You should do this for a type check instead:`if _, ok := m.(*cardinality); ok {`

But.... this doesn't really apply, since cardinality should be unique, I'm not sure this shortcut works any more.

snikch reviewed (changes requested):
A few changes to be made.

I have to apologise for not leading by example with the comments. I dropped the ball in the measurers. It'd be good to add some comments in the appropriate format for each of the measurers.

mattevans committed:
Removed shortcut to bypass non-metricable cell for cardinality. Now passing measuer{} to IsMetricable to determine if metric can run

mattevans committed:
IsMetricable() now requires measurer{}

mattevans committed:
Comments. Implemented cardinality correctly. Added valueCount metric

mattevans committed:
Tests updated to reflect changes made

mattevans committed:
Using decimal math for median metric

mattevans committed:
Commenting

mattevans committed:
Comments

mattevans commented:
Thanks @snikch - Really appreciate it.

- Added commenting
- Fixed smaller issues (decimal math, removed extra loop, etc)
- `cardinality` metric now actually calculates the cardinality. 👍 
- Added `valueCount` metric, which is exactly that.

---

**Questions on `mode` metric:**

> Why do we return an empty loop if the modes are all of the numbers / one? I haven't dealt with modes too much but wouldn't it still be relevant to have the only possible value returned regardless?

I was always under the impression mode was the value(s) that occur/repeat most often within a set. If all values have the same number of occurrences, then they aren't technically a mode? After a quick google, some say they should be considered mode, others say they shouldn't. Now I'm not so sure. hahaha. Thoughts? I'll ask Lance if he has a preference for either option.

**Examples:**
```
[1,1,2,3,4,5,5,5]
Mode = 5
```
```
[1,1,1,2,3,3,3]
Mode = [1,3]
```
```
[1,1,1,2,2,2,3,3,3]
Mode = nil
```
```
[1,2,3,4,5]
Mode = nil
```

If the dataset is `one` (referenced as `tip` in the code), then it means no repeating values were found. Now that I've added comments (and cleaned it up), it should make a bit more sense. Sorry about that.

> In fact we should quickly explain what each of the averages are - I always get them mixed up, so it'd be nice not to have to think too hard.

Added comments to each metric.

---

**Metricable string values**

With `cardinality` (and now `valueCount`) measurers, these needed to be run against `StringCells`. You pointed out the shortcut wasn't going to work once cardinality was implemented correctly. That shortcut is gone!

To resolve, I'm passing the `measurer{}` to `IsMetricable()` ([see here](https://github.com/snikch/aggro/pull/3/files#diff-c432882c123463117b32161d81b1b461L183)).

Previously, the type of cell determined if the field was metricable. Passing the measurer, allows us to run certain metrics against certain cells. As it stands, only needed for `StringCells`, but could be handy for different metrics in the future. 

Do you think this is an alright solution? Maybe I should define what measurers can be used on what cells outside of each XCell implementation?

mattevans commented:
Going to merge this after our discussion/look over last night. 
Can make further adjustments if needed.

Hope your Xmas party is in full swing! 

![3880216](https://cloud.githubusercontent.com/assets/714721/21206891/a20656a0-c2c9-11e6-93d3-64e58a22056a.jpg)
